### PR TITLE
Tools: auto fetch parameter if setting an _ENABLE one

### PR DESCRIPTION
**Pull request by khancyr**

### CODE REVIEW

**Comment by Pedals2Paddles** [comment]
I think this would actually be something good for @meee1 to implement in the GCS too.

**Comment by WickedShell** [comment]
@Pedals2Paddles there are actually more enable parameters then just ones that have ENABLE in the name. The good news is that when you set an enable parameter the acknowledgement tells you the total count has changed and the GCS can automatically re fetch the parameters. MAVProxy doesn't do this correctly at the moment which is why this type of patch is showing up in autotest. MP and others though should simply be watching the count and downloading a new parameter list when it changes.

**Comment by magicrub** [comment]
This is a good start. These also need it:
PLND_ENABLED
ARSPD_TYPE
ARSPD2_TYPE
BATT_MONITOR
BATT2_MONITOR
CAN_P1_DRIVER
COMPASS_PMOT_EN
OSD_TYPE
CHUTE_ENABLED
RSSI_TYPE
TCALENABLED <--- this is missing the _ after TCAL
WENC_TYPE

**Comment by amilcarlucas** [comment]
I vote for
s/PLND_ENABLED/PLND_ENABLE
s/CHUTE_ENABLED/CHUTE_ENABLE
s/TCALENABLED/TCAL_ENABLE
Before the 3.6.0 release

**Comment by khancyr** [comment]
this could correct the problem : https://github.com/ArduPilot/MAVProxy/pull/536

**Comment by peterbarker** [comment]
Globbing on `*_ENABLE?` and `*_EN`would cut this down and make it less likely to atrophy.

I think we should put this in - but then fix MAVProxy and remove it!

**Comment by khancyr** [comment]
@peterbarker, I made a fix for mavproxy.
The commit title is wrong... as some parameter with _TYPE are enable type... my solution is just a dirty patch to prevent autotest to fail until we work on something better. I remember a solution proposed by QGC guys but don't remember it or where it was wrotten...

**Comment by peterbarker** [comment]
On Sun, 26 Aug 2018, Pierre Kancir wrote:

> @peterbarker, I made a fix for mavproxy.
> The commit title is wrong... as some parameter with _TYPE are enable type... my solution is just a dirty patch to prevent autotest to fail until we work
> on something better. I remember a solution proposed by QGC guys but don't remember it or where it was wrotten...

Yes, I realise globbing for _ENABLE and _ENABLED won't get everything. 
But I *very* much doubt there will be false-positives.

So something like:

def should_fetch_all_for_parameter_change(self, foo):
     if fnmatch(foo, "*_ENABLE") or fnmatch(foo, "*_ENABLED"):
         return True
     if foo in [ ... ]:
         return True
     return False

**Comment by khancyr** [comment]
@peterbarker Done !

**Comment by peterbarker** [comment]
Merged, thanks!  Hopefully we can sort something out with MAVProxy - but even then we'll have a race condition we'll need to resolve as MAVProxy wanders off to fetch parameters while we try to set things it hasn't heard about yet :-/

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -492,6 +508,8 @@ def set_parameter(self, name, value, add_to_context=True):
                 # yes, exactly equal.
                 if add_to_context:
                     self.context_get().parameters.append((name, old_value))
+                if self.should_fetch_all_for_parameter_change(name.upper()) and value == 1:
```
Comment (amilcarlucas) [inline Tools/autotest/common.py]: For *_TYPE it makes sense to use >= 1

**Pull request merged**